skfp/utils/splitters.py

```diff
@@ -0,0 +1,239 @@
+from typing import Union
+
+import numpy as np
+from rdkit.Chem import Mol
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
+from skfp.utils.validators import ensure_mols
+
+
+def scaffold_train_test_split(
+    data: Sequence[Union[str, Mol]],
+    train_size: float = 0.8,
+    test_size: float = 0.2,
+    return_indices: bool = False,
+    include_chirality: bool = False,
+) -> Union[
+    tuple[list[str], list[str]],
+    tuple[list[Mol], list[Mol]],
+    tuple[list[int], list[int]],
+]:
+    """
+    Split a list of SMILES or RDKit 'Mol' objects into train and test subsets using Bemis-Murcko [1]_ scaffolds.
+
+    This function ensures that similar molecules (sharing the same scaffold) are in the same split.
+    This method, suggested in the MoleculeNet [2]_ paper, helps to test the model's ability to
+    generalize to entirely new scaffolds.
+
+    The split is fully deterministic, with the smallest scaffold sets assigned to the test
+    subset and the largest to the training subset.
+
+    The split fractions (train_size, test_size) must sum to 1.
+
+    Parameters
+    ----------
+    data : sequence
+        Sequence representing either SMILES strings or RDKit 'Mol' objects.
+
+    train_size : float, default=0.8
+        The fraction of data to be used for the train sub.
+
+    test_size : float, default=0.2
+        The fraction of data to be used for the test sub.
+
+    return_indices : bool, default=False
+        Whether the method should return the input object subsets, i.e. SMILES strings
+        or RDKit `Mol` objects, or only the indices of the subsets.
+
+    include_chirality: bool, default=False
+        Whether to take chirality of molecules into consideration.
+
+    Returns
+    ----------
+    subsets : tuple[list, list]
+        A tuple of train and test subsets. The format depends on `return_indices`:
+        - if `return_indices` is False, returns lists of SMILES strings
+        or RDKit 'Mol' objects depending on the input
+        - if `return_indices` is True, returns lists of indices
+
+    References
+    ----------
+    .. [1] Bemis, G. W., & Murcko, M. A. (1996). The properties of known drugs. 1. Molecular frameworks.
+    Journal of Medicinal Chemistry, 39(15), 2887-2893.
+    .. [2] Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande (2017).
+    MoleculeNet: A Benchmark for Molecular Machine Learning. Chemical Science, 9(2), 513-530.
+    """
+    np.testing.assert_almost_equal(train_size + test_size, 1.0)
+
+    scaffolds = _create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_cutoff: int = int(train_size * len(data))
+    train_ids: list[int] = []
+    test_ids: list[int] = []
+
+    for scaffold_set in scaffold_sets:
+        if len(train_ids) < train_cutoff:
+            train_ids.extend(scaffold_set)
+        else:
+            test_ids.extend(scaffold_set)
```

Comment:
You have the scaffold sets sorted from smallest to largest. So I would write this as:
```
desired_test_size = int(test_size * len(data))

for scaffold_set in scaffold_sets:
    if len(test_ids) < desired_test_size:
        test_ids.extend(scaffold_set)
    else:
        train_ids.extend(scaffold_set)
```

Reply:
This would also follow the descriptions from papers more explicitly - we *first* construct the test set from smallest scaffold sets, and *then* put the rest into train set